cmd/risor-modgen/main.go

```diff
@@ -0,0 +1,332 @@
+		if !entry.IsDir() {
+			continue
+		}
+		path := filepath.Join(options.Modules, entry.Name())
+		if err := runInDir(path, options); err != nil {
+			return fmt.Errorf("module %q: %w", path, err)
+		}
+	}
+	return nil
+}
+
+func runInDir(dir string, options Options) error {
+	fset := token.NewFileSet()
+	ignoreRegex, err := regexp.Compile(options.IgnoreFiles)
+	if err != nil {
+		return fmt.Errorf("parse -ignore flag: %w", err)
+	}
+	pkgs, err := parser.ParseDir(fset, dir, func(fi fs.FileInfo) bool {
+		return !strings.HasSuffix(fi.Name(), "_gen.go") &&
+			!strings.HasSuffix(fi.Name(), "_test.go") &&
+			!strings.HasSuffix(fi.Name(), "_example.go") &&
+			!ignoreRegex.MatchString(fi.Name())
+	}, parser.ParseComments|parser.DeclarationErrors)
+
+	if err != nil {
+		return err
+	}
+
+	if len(pkgs) != 1 {
+		var pkgNames []string
+		for name := range pkgs {
+			pkgNames = append(pkgNames, name)
+		}
+		return fmt.Errorf("directory must only contain 1 package, but got: %s", pkgNames)
+	}
+
+	for _, pkg := range pkgs {
+		mod, err := Parse(fset, pkg)
+		if err != nil {
+			return err
+		}
+
+		if !mod.HasGenerateComment {
+			fmt.Printf("Skipping, missing risor:generate comment: %s\n", dir)
+			continue
+		}
+
+		genFile := filepath.Join(dir, mod.Name+"_gen.go")
+
+		oldHash := hashFile(genFile)
+		if err := mod.WriteFile(genFile, options); err != nil {
+			return fmt.Errorf("write generated file: %w", err)
+		}
+		newHash := hashFile(genFile)
+
+		if oldHash == newHash {
+			fmt.Printf("No changes to file: %s\n", genFile)
+			continue
+		}
+
+		stat, err := os.Stat(genFile)
+		if err != nil {
+			return fmt.Errorf("stat generated file: %w", err)
+		}
+		fmt.Printf("Wrote to file: %s (%d B)\n", genFile, stat.Size())
+	}
+	return nil
+}
+
+func hashFile(path string) string {
+	hash := sha512.New()
+	file, err := os.Open(path)
+	if err != nil {
+		return ""
+	}
+	defer file.Close()
+
+	if _, err := io.Copy(hash, file); err != nil {
+		return ""
+	}
+	return fmt.Sprintf("%x", hash.Sum(nil))
+}
+
+type Module struct {
+	Name string
+
+	HasGenerateComment bool
+	skipModulesFunc    bool
+
+	fset             *token.FileSet
+	buildConstraints []string
+	exportedFuncs    []ExportedFunc
+}
+
+func Parse(fset *token.FileSet, pkg *ast.Package) (*Module, error) {
+	mod := &Module{
+		Name: pkg.Name,
+		fset: fset,
+	}
+	for path, file := range pkg.Files {
+		if err := mod.parseFile(file); err != nil {
+			return nil, fmt.Errorf("file %q: %w", path, err)
+		}
+	}
+	return mod, nil
+}
+
+func (m *Module) WriteFile(path string, options Options) error {
+	file, err := os.Create(path)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	m.Fprint(file, options)
+	return nil
+}
+
+var tmpl = template.Must(template.New("generated").Parse(`// Code generated by risor-modgen. DO NOT EDIT.
+{{- if .BuildConstraints }}
+{{ range .BuildConstraints }}
+{{ . }}
+{{- end }}
+{{- end }}
+
+package {{ .Package }}
+
+import (
+	"context"
+	"math"
```

Comment:
Why import this if it's not being used?

Reply:
It's used if you have for example an int parameter. The object.AsInt returns a int64, so to convert to int we first need to check if the value is not bigger than math.MaxInt

Reply:
I added import tracking. So now the `Stop compile error` section is gone.